src/main/java/org/folio/rest/impl/CheckinReceivePiecesHelper.java

```diff
@@ -186,43 +178,31 @@ private void checkIfAllPiecesFound(List<String> pieceIds, List<Piece> pieces) {
     }
   }
 
-  /**
-   * Updates items in the inventory storage with receiving/check-in details if any. On
-   * success updates corresponding records as received
-   *
-   * @return {@link CompletableFuture} which holds map with PO line id as key
-   *         and list of corresponding pieces as value
-   */
-  CompletableFuture<Map<String, List<Piece>>> updateInventoryItems(Map<String, List<Piece>> piecesGroupedByPoLine) {
-    // Collect all piece records with non-empty item ids. The result is a map
-    // with item id as a key and piece record as a value
-    Map<String, Piece> piecesWithItems = collectPiecesWithItemId(piecesGroupedByPoLine);
+  private PoLine searchPoLineById(List<PoLine> poLines, String polineId) {
+    for (PoLine  poline: poLines){
+      if (poline.getId().equals(polineId)){
+        return poline;
+      }
+    } return null;
+  }
 
-    // If there are no pieces with ItemId, continue
-    if (piecesWithItems.isEmpty()) {
-      return completedFuture(piecesGroupedByPoLine);
+  private CompletableFuture<Boolean> updateHoldingAndItem(JsonObject item, Piece piece, PoLine poLine, String locationId) {
+    if (holdingUpdateOnCheckinReceiveRequired(piece, locationId, poLine)) {
+      return inventoryHelper.getOrCreateHoldingsRecord(poLine.getInstanceId(), locationId)
+        .thenCompose(holdingId -> {
+          item.put("holdingsRecordId", holdingId);
+          return receiveInventoryItemAndUpdatePiece(item, piece);
+        })
+        .exceptionally(t -> false);
```

Comment:
Please add error if exception happens like it is handled in `receiveInventoryItemAndUpdatePiece`. Otherwise in case there is an issue handling holding, the response will not take into account it and would consider processed piece as "success"

Reply:
updated error handling